### DISCUSSION TITLE
Does someone know how to debug Compiler Plugins?

### DISCUSSION THREAD

**Discussion by orchestr7** [akuleshov7/ktoml#132] [Q&A] [answered]
I am trying to debug tests from https://github.com/akuleshov7/ktoml/pull/125/files but do not know how to debug core logic (from kotlin.serialization compiler plugin) in IDEA.

1) I tried to run: `gradle :ktoml-core:cleanJvmTest :ktoml-core:jvmTest --tests "com.akuleshov7.ktoml.decoders.InlineDecoderTest" --no-daemon -i -Dorg.gradle.debug=true -Dkotlin.compiler.execution.strategy=in-process`

2) I put breakpoints to the test, ktoml and serialization core

3) I attached remote debugger in IDEA to the process

4) It connected, but **missed all break points**. Test simply failed. What I am doing wrong?

@bishiboosh @Peanuuutz may be you had some experience with that?

**Comment by eav-eav-eav** [comment]
Well...talk about what is concerned?

**Comment by bishiboosh** [comment]
(I took the initiative to move that to a discussion in order not to introduce an issue where there is not really one)

I found that https://github.com/Foso/MpApt/wiki/How-to-debug-Kotlin-Compiler-Plugin#kotlin-jvmjs, maybe that could help you ?

**Reply by orchestr7** [reply]
Yeah, I have seen that one. That is used to debug compiler plugins during the `build` phase (when IR is generated by the compiler).

But I would like to debug, for example, the method decodeFromString() in unit tests on runtime. Right now I am not able to see the stack to the `serialization-core`. Debugger only points to `@Serializable` annotation without any stack to real methods provided by serialization library.

**Reply by orchestr7** [reply]
I am trying to find out how we can debug Ktoml and get the full stack to the methods provided by serialization-core. 

For example to see the stack from serialisation-core to our library in this case: https://github.com/akuleshov7/ktoml/issues/124

**Reply by bishiboosh** [reply]
hmmm I see what you mean, debugging the output provided by a compiler plugin...Maybe you should ask that on the kotlin forums, I don't really have a clue right now :/

**Comment by orchestr7** [comment, marked as answer]
@bishiboosh FYI: Leonid, an author of kotlinx.serialization has given an awesome answer [here](https://github.com/Kotlin/kotlinx.serialization/issues/1888#issuecomment-1080544035)

At least this can help. This is how primitive decoding look like, I love this code. Finally I have understood the logic of generated code:

```
@NotNull
public InlineDecoderTest.Color deserialize(@NotNull Decoder decoder) {
  Intrinsics.checkNotNullParameter(decoder, "decoder");
  SerialDescriptor var2 = this.getDescriptor();
  boolean var3 = true;
  int var5 = 0;
  long var6 = 0L;
  CompositeDecoder var8 = decoder.beginStructure(var2);
  if (var8.decodeSequentially()) {
     var6 = var8.decodeLongElement(var2, 0);
     var5 |= 1;
  } else {
     while(var3) {
        int var4 = var8.decodeElementIndex(var2);
        switch(var4) {
        case -1:
           var3 = false;
           break;
        case 0:
           var6 = var8.decodeLongElement(var2, 0);
           var5 |= 1;
           break;
        default:
           throw new UnknownFieldException(var4);
        }
     }
  }

  var8.endStructure(var2);
  return new InlineDecoderTest.Color(var5, var6, (SerializationConstructorMarker)null);
```
